src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -308,9 +309,32 @@ internal BoundStatement BindPossibleEmbeddedStatement(StatementSyntax node, Diag
                     diagnostics.Add(ErrorCode.ERR_BadEmbeddedStmt, node.GetLocation());
 
                     // fall through
-                    goto case SyntaxKind.ExpressionStatement;
+                    goto case SyntaxKind.LockStatement;
 
                 case SyntaxKind.ExpressionStatement:
+                    // In Scripts, we allow statements to not have semicolons.  However, that only
+                    // applies to non-embedded statements.  So check if we haven't emitted an error
+                    // for that case and do that here.  We don't do this in the parser for
+                    // simplicity of that impl and to prevent unnecessary helper methods in a hot
+                    // path where we're trying to keep the number of stack frames we have low.
+                    if (node.SyntaxTree.Options.Kind == SourceCodeKind.Script)
+                    {
+                        var expressionStatement = (ExpressionStatementSyntax)node;
+                        var semicolonToken = expressionStatement.SemicolonToken;
+
+                        // Do not add a new error if the same error was already added.
+                        if (semicolonToken.IsMissing &&
+                            !semicolonToken.GetDiagnostics().Contains(diagnosticInfo => (ErrorCode)diagnosticInfo.Code == ErrorCode.ERR_SemicolonExpected))
+                        {
+                            diagnostics.Add(
+                                ErrorCode.ERR_SemicolonExpected,
+                                Location.Create(node.SyntaxTree,
+                                    new TextSpan(expressionStatement.Expression.GetLastToken().Span.End, 0)));
```

Comment:
Parse errors should be reported from the parsing phase.

Reply:
That's not actually true.  There are many of these sorts of erorrs we do not report from the parsing phase.  That's because it's actually not a good thing to report errors during parsing for nodes that fit fine into our syntactic model.  For one thing, it prevents incrementality, for another it represents a concept outside the bounds of what the parser is for.  i.e. the parser's job is to say: let me fit thsi text to this model.  If the text can fit, then there is no issue at its level.  So the only issues it really needs to report are "i needed a token, and it wasn't there." or "i ran into a token i couldn't deal with".

It should not actually report cases of "i didn't need a token and i didn't run into one".  If we have normative rules on top of our existing syntactic model (i.e. ExpressionStatements require semicolons in some contexts, but not in others), then doing that at a later phase is much cleaner.

Reply:
Note: a whole bunch of changes went in several years back to try to get the parser more on-track in this regard.  Lots of errors have moved out to better phases (i.e. the decl table or binding).  It's def the case that those errors *could* be pushed down to the parser, but it's much cleaner to not have the parser do *contextual* parsing like this, and leave context to other phases.

Reply:
Reporting a syntax error for something as fundamental as a missing semicolon belongs in the parser.

---
In reply to: [378511480](https://github.com/dotnet/roslyn/pull/41589#discussion_r378511480) [](ancestors = 378511480)

Reply:
I don't know what makes it "fundamental".  The syntax model legitimately says that it is optional.

There are *many* sorts of similar changes we pulled out of the parser over the years.  For example, it's completely bogus to say `namespace A::B`.  but we don't report that in the parser, but later on in decl trees.  Similar to issues around 'await' and whatnot.

If it helps, think of this error message as basically saying "Error: simple expressions are not allowed in embedded locations in script files".  just that it also says what a simple solution is to that by "adding a semicolon".

--

Note: this also doesn't address the important points i made about incrementality.  It's *not* a virtue to have unnecessary errors during parsing.  It actually makes it harder to reuse nodes and it bubbles up all over the place as errors that block other behaviors.  This is because higher layers have to take the conservative position that a parse error means things went bad and we don't necessarily know what's going on.  That's not the case here.  We know exactly what's happening, and it's totally trivial to support.  

Finally, i can try to undo this and go another route.  But it would seem super unfortunate to me.  This check honestly does not belong in the parser to me.  Just as we error *later* if you do something like `if (foo) var x = 1;` (because a local decl is not allowed here), we shoudl error later if you use a naked-script-expr here.